libs/pageserver_api/src/models.rs

```diff
@@ -841,36 +841,36 @@ impl TryFrom<u8> for PagestreamBeMessageTag {
 
 #[derive(Debug, PartialEq, Eq)]
 pub struct PagestreamExistsRequest {
-    pub latest: bool,
+    pub horizon: Lsn,
```

Comment:
I think the semantics would be clearer if we used a stronger type for the horizon and didn't rely on the `Lsn::MAX` and `Lsn::INVALID` values or replaced `horizon` and `lsn` with a `Range<Lsn>`.

Something like:

```
enum LsnHorizon {
  None,
  At(Lsn),
  Infinite
}
```

Reply:
While writing this comment I realised that I don't get why we need the `Lsn::INVALID` value for horizon. `horizon == Lsn::INVALID` should be equivalent to `latest=false`, so could we set horizon to the request lsn in these cases instead?

I think Arseny had a similar comment.

Reply:
Well, introducing new typer requires too much refactoring and frankly speaking I do not some significant benefits of presence of such type. 
Also it will not allow to use very simple and expressive formula which I really like:
```
let effective_lsn = Lsn::max(lsn, Lsn::min(request_horizon, last_record_lsn));
```

Concerning second question: yes it is not srtictly needed to pass Lsn::INVALID. It is just more convenient. Both in page server tests and in neon extension where decision whether to use latest version or not usually taken in the context where request lsn is not known.

Reply:
> Well, introducing new typer requires too much refactoring and frankly speaking I do not some significant benefits of presence of such type. Also it will not allow to use very simple and expressive formula which I really like:
> 
> ```
> let effective_lsn = Lsn::max(lsn, Lsn::min(request_horizon, last_record_lsn));
> ```

That's very unclear to me. What is `request_horizon` ? Is somehow related to server-side GC horizon perhaps? Reading this, I keep getting confused what exactly is horizon, what the special LSN values mean, etc.